yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts

```diff
@@ -0,0 +1,210 @@
+   * pending indexes already stored from previous syncs.
+   * @param preTags - The pre-tags containing the directional app tagging secrets and the indexes that are to be
+   * stored in the db.
+   * @param txHash - The hash of the pending tx that used the given pre-tags to compute private log tags.
+   * @throws If any two pre-tags contain the same directional app tagging secret. This is enforced because we care
+   * only about the highest index for a given secret that was used in the tx. Hence this check is a good way to catch
+   * bugs.
+   * @throws If a secret + txHash pair already exists in the db with a different index value. It should never happen
+   * that we would attempt to store a different index for a given secret-txHash pair because we always store just the
+   * highest index for a given secret-txHash pair. Hence this is a good way to catch bugs.
+   * @throws If the newly stored pending index is lower than or equal to the last finalized index for the same secret.
+   * This is enforced because this should never happen if the syncing is done correctly as we look for logs from higher
+   * indexes than finalized ones.
+   */
+  async storePendingIndexes(preTags: PreTag[], txHash: TxHash) {
+    // The secrets in pre-tags should be unique because we always store just the highest index per given secret-txHash
+    // pair. Below we check that this is the case.
+    const secretsSet = new Set(preTags.map(preTag => preTag.secret.toString()));
+    if (secretsSet.size !== preTags.length) {
+      throw new Error(`Duplicate secrets found when storing pending indexes`);
+    }
+
+    for (const { secret, index } of preTags) {
+      const secretStr = secret.toString();
+      const existing = (await this.#pendingIndexes.getAsync(secretStr)) ?? [];
+
+      // Throw if the new pending index is lower than or equal to the last finalized index
+      const lastFinalizedIndex = await this.#lastFinalizedIndexes.getAsync(secretStr);
+      if (lastFinalizedIndex !== undefined && index <= lastFinalizedIndex) {
+        throw new Error(
+          `Cannot store pending index ${index} for secret ${secretStr}: ` +
+            `it is lower than or equal to the last finalized index ${lastFinalizedIndex}`,
+        );
+      }
+
+      // Check if this secret + txHash combination already exists
+      const existingEntry = existing.find(entry => entry.txHash === txHash.toString());
+
+      if (existingEntry) {
+        // If it exists with a different index, throw an error
+        if (existingEntry.index !== index) {
+          throw new Error(
+            `Cannot store index ${index} for secret ${secretStr} and txHash ${txHash.toString()}: ` +
+              `a different index ${existingEntry.index} already exists for this secret-txHash pair`,
+          );
+        }
+        // If it exists with the same index, ignore the update (no-op)
```

Comment:
What if it existed with a lower index? This seems to rely on the caller always passing the largest index, which seems maybe not great as that is greatly affecting the internal logic here. I'd strongly consider reworking that, or taking the highest index in here instead of outside, making the caller logic simpler.